### ISSUE TITLE
Audible support

### ISSUE THREAD

**Issue by nullbasis** [ytdl-org/youtube-dl#28317]
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running youtube-dl version **2021.03.02**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- Single book: https://www.audible.com/pd/The-Invasion-Audiobook/1338641158
- Single book: https://www.audible.com/webplayer?asin=1338641158
- Playlist: N/A

## Description

Username and password would be required to download books. Since Audible only supports listening to books in the browser or in an app, it would be incredible if yt-dl could download one's purchases.

I'm not sure which of the above example links would be better as input. The latter is the page where the book plays directly, but the former is more obviously accessible to the average user.

**Comment by afe2c3** [comment]
This seems like a good idea, I +1 this. However, development will stall since Audible is paid.

**Comment by nullbasis** [comment]
Honestly, if it would help, I'd be willing to buy and gift some cheap book to someone who would do the dev work.

**Comment by babbittw** [comment]
I use Audible a good amount. I'll take a shot at it.

**Comment by DerZyklop** [comment]
Maybe interesting here: Some people who know how to find a specific network request with devTools downloaded stuff with youtube-dl. 

https://forum.videohelp.com/threads/396514-Downloading-audio-from-stories-audible-com

They wrote that the links to audio files expire qickly (~60 seconds).

I did not have any luck with that, because i had an Error 403 no matter how fast i did it.

**Comment by 89z** [comment]
The Web client uses DRM, but it seems the Android client is totally open:

http://nbcnews-injected.simplecastaudio.com/ae183fe2-c634-458a-93dd-5770f0676f77/episodes/8283d7a6-7138-4bce-b173-f304ee245234/audio/128/default.mp3/default.mp3_6e730a9efdbbca96d5932358fb3b272b_31363304.mp3

This is from a free item, not sure about other items. If people are interested, I can see about working up an API.

**Comment by nullbasis** [comment]
I also keep getting 403 errors with the method in that forum post. Amazon must have "fixed" things since then.

89z, an API would be great!

**Comment by 89z** [comment]
please post issue on my tracker, and I will work on it https://github.com/89z/mech

**Comment by 89z** [comment]
https://github.com/89z/mech/issues/14

**Comment by LinusU** [comment]
This could probable help:

https://github.com/openaudible/openaudible/blob/34a305bd384c1918b3ef12b917aab53aed30aa74/src/main/java/org/openaudible/download/DownloadJob.java#L70-L84

For me, clicking "Download" on the Audible site is basically useless unless my internet connection is super stable. Since I'm currently out traveling in Asia it never is, and thus I have been having a really hard time getting my books. I manage to use `curl` with the continue-flag to resume my transfer, and trying repeatedly seems to make progress!

```sh
curl \
  -C - \
  -o my-book.aax \
  "https://cds.audible.com/download?user_id=$USERID&product_id=$SKU&codec=LC_128_44100_Stereo&awtype=AAX&cust_id=$USERID" \
  --header 'User-Agent: Audible ADM 6.6.0.19;Windows Vista  Build 9200'
```

I managed to find my `$USERID` by having the developer console open whilst trying to download the file from the Audible site, and the `$SKU` can be found in the source for the page of the book e.g:

<img width="692" alt="Screenshot 2023-03-16 at 11 05 54" src="https://user-images.githubusercontent.com/189580/225511681-76632cca-30cf-4473-8228-17bc69861108.png">

**Comment by LinusU** [comment]
I just published a little tool that downloads, and most importantly, retries on each lost connection 🎉

**Comment by dirkf** [comment]
The tool implements https://github.com/ytdl-org/youtube-dl/issues/28317#issuecomment-1471274712.

As this solution is to populate
```
https://cds.audible.com/download?user_id={USERID}&product_id={SKU}&codec=LC_128_44100_Stereo&awtype=AAX&cust_id={USERID}
```
an extractor has to find `USERID` and `SKU`. The regional page  https://www.audible.co.uk/pd/The-Invasion-Audiobook/1338641158, non-JS, not logged in, has the `SKU` value in a JS assignment matching `"sku": "(.+)"`.

Possibly if logged-in cookies were sent the `.member` of the JS object would also be set with the `USERID` value.

For a URL like https://www.audible.com/webplayer?asin=1338641158, the `SKU` isn't available: perhaps rewrite the URL using the regional domain: apparently even this works:
```
https://www.audible.co.uk/pd/whatever-Audiobook/1338641158
```

**Comment by 4cq2** [comment]
this appears to download AAX files, which are encrypted:

https://wikipedia.org/wiki/Audible_(service)#Quality

if thats the case, I believe that is explicitly an anti-feature of YouTube-DL. @LinusU, do you have some reason to believe these are unencrypted files?

**Comment by 4cq2** [comment]
https://github.com/LinusU/audible-dl/issues/1

**Comment by LinusU** [comment]
@4cq2 yes, they are encrypted. I should probably say that I just threw this tool together quickly to solve my particular problem right now: I can't download this specific audiobook on the internet connection I'm on right now.

It's clunky to use for sure since you need to know your customer_id, and the sku of the book.

In order to actually listen to the book, I'm using the instructions here: https://gist.github.com/r15ch13/0c548be006431607bf1eaecefdc0591a?permalink_comment_id=2800421#gistcomment-2800421

I once extracted my activation bytes from a Windows computer and have saved it in 1Password, so for me personally this is really easy.

-----

It would be great with a PR to this repo that does all of the things: 1) variable extraction from url + login or cookies, 2) download the AAX file, and finally 3) decrypt it using ffmpeg. Ultimately, I only had an hour to spare on this, and I like Rust very much, so I did a separat tool that only solved my problem...

-----

> Possibly if logged-in cookies were sent the .member of the JS object would also be set with the USERID value.

@dirkf this is very interesting! Unfortunately, it doesn't seem to work for me though, `member` is an empty object in the source served to me when logged in. Maybe this has something to do with my membership not being active though 🤔 (that is, I'm not currently on an active plan)

![Screenshot 2023-03-18 at 10 48 26](https://user-images.githubusercontent.com/189580/226083495-1a63c6e8-ddd6-45b9-a90e-5005bdd1098c.png)

(e.g. https://www.audible.com/pd/The-Hobbit-Dramatised-Audiobook/B002VA91WK)

**Comment by 4cq2** [comment]
@LinusU decrypting media is typically an anti-feature of YouTube-DL. which means that any issues or pull requests discussing that are usually closed on those grounds. The line is somewhat blurry, as stuff like Widevine is explicitly not allowed, while other stuff like HLSe is allowed. this seems to pull a static key from the registry, which seems like it might be OK, in regards to the nebulous YouTube-DL rules.

YouTube-DL already can accept an argument that gets passed to FFmpeg, so that would apply here with `-activation_bytes`. I dont think it should be YouTube-DL job to go digging into a users registry, just like I dont think it should be YouTube-DL job to go digging into a users cookies, but I guess that ship has sailed, so who knows. I will be interested to see how a maintainer views this issue.

**Comment by dirkf** [comment]
If all that's needed is to pass a "password" available to a properly authenticated user to _ffmpeg_, that's fine. The "live reenactment" shown [here](https://www.da.vidbuchanan.co.uk/blog/netflix-on-asahi.html) won't happen.

**Comment by 4cq2** [comment]
> The "live reenactment" shown [here](https://www.da.vidbuchanan.co.uk/blog/netflix-on-asahi.html) won't happen.

with version L3 (as detailed in that article), the CDM is not "looked inside of", but rather the `ClientId` bytes are used as part of the license request/response process. However, I get your point.

**Comment by slavakurilyak** [comment]
For inspiration, check out https://github.com/rmcrackan/Libation

**Comment by dirkf** [comment]
Less obscure inspiration (not encrypted into C#, eg) is available by searching the Web for "ffmpeg aax", or just re-reading the issue.

_yt-dl_ is not going to include any cracking tool, but a user who has identified the correct value to be passed to _ffmpeg_ as `-activation_bytes ....` can pass that option in a command run with `-exec ...`.

It may be possible to pass any such parameters to a downloader or postprocessor step that uses _ffmpeg_ instead, if the parameters are inserted in the correct position in _yt-dl_'s constructed _ffmpeg_ command (or in _yt-dlp_ which allows for controlling the position in a postprocessor command).
